src/dag/write.rs

```diff
@@ -1,55 +1,195 @@
-use super::chunk::Chunk;
 use super::key::Key;
+use super::{chunk::Chunk, meta_generated::meta};
 use super::{read, Result};
 use crate::kv;
+use async_recursion::async_recursion;
+use async_std::sync::RwLock;
+use futures::future::try_join_all;
+use futures::try_join;
+use std::collections::HashSet;
+
+#[derive(Debug, Default)]
+struct HeadChange(Option<String>, Option<String>);
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+enum RefCount {
+    Inc = 1,
+    Dec = -1,
+}
 
 pub struct Write<'a> {
     kvw: Box<dyn kv::Write + 'a>,
+    changed_heads: RwLock<Vec<HeadChange>>,
+    mutated_chunks: RwLock<HashSet<String>>,
 }
 
 impl<'a> Write<'_> {
     pub fn new(kvw: Box<dyn kv::Write + 'a>) -> Write {
-        Write { kvw }
+        Write {
+            kvw,
+            changed_heads: Default::default(),
+            mutated_chunks: Default::default(),
+        }
     }
 
     pub fn read(&self) -> read::Read {
         read::Read::new(self.kvw.as_read())
     }
 
     pub async fn put_chunk(&mut self, c: &Chunk) -> Result<()> {
+        // TODO: These can be done in parallel.
+
         self.kvw
             .put(&Key::ChunkData(c.hash()).to_string(), c.data())
             .await?;
+
         if let Some(meta) = c.meta() {
             self.kvw
                 .put(&Key::ChunkMeta(c.hash()).to_string(), meta)
                 .await?;
         }
+
+        self.mutated_chunks
+            .write()
+            .await
+            .insert(c.hash().to_string());
+
         Ok(())
     }
 
     pub async fn set_head(&mut self, name: &str, hash: Option<&str>) -> Result<()> {
+        let old_hash = self.read().get_head(name).await?;
+
+        // TODO: These can be done in parallel.
+
+        let head_key = Key::Head(name).to_string();
         match hash {
-            None => Ok(self.kvw.del(&Key::Head(name).to_string()).await?),
-            Some(h) => Ok(self
-                .kvw
-                .put(&Key::Head(name).to_string(), h.as_bytes())
-                .await?),
+            None => self.kvw.del(&head_key).await?,
+            Some(h) => self.kvw.put(&head_key, h.as_bytes()).await?,
         }
+
+        self.changed_heads
+            .write()
+            .await
+            .push(HeadChange(hash.map(str::to_string), old_hash));
+
+        Ok(())
     }
 
     pub async fn commit(self) -> Result<()> {
+        self.collect_garbage().await?;
         Ok(self.kvw.commit().await?)
     }
 
     #[allow(dead_code)]
     pub async fn rollback(self) -> Result<()> {
         Ok(self.kvw.rollback().await?)
     }
+
+    async fn set_ref_count(&self, hash: &str, count: u16) -> Result<()> {
+        // Ref count is represented as a u16 stored as 2 bytes using BE.
```

Comment:
[question] did you consider using flatbuffers for the single value? Kinda silly but (a) it takes care of the serialization for you, and (b) is extensible (not sure what we'd use that for but you never know).

Reply:
I did consider it but feels like a lot of code for storing a single number

Reply:
Agree. I believe every platform we run on is LE. Is there any value to sticking with that?

Reply:
no preference between LE or BE. I just picked one... I'll change to LE for simplicity